File: app/src/main/java/com/artamonov/millionplanets/sectors/SectorsYouDialog.java
package com.artamonov.millionplanets.sectors;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.NumberPicker;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDialogFragment;
import com.artamonov.millionplanets.R;
import com.artamonov.millionplanets.model.ObjectModel;
import com.artamonov.millionplanets.model.User;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.Transaction;
import java.util.List;

public class SectorsYouDialog extends AppCompatDialogFragment {
    FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
    private List<User> userList;
    private List<ObjectModel> objectModelList;
    private NumberPicker numberPicker;
    private FirebaseUser firebaseUser;
    private DocumentReference documentReference;
    private DocumentReference documentReferenceUser;
    private DocumentReference documentReferenceInventory;
    private boolean isPlanetTab;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
        documentReferenceInventory =
                firebaseFirestore.collection("Inventory").document(firebaseUser.getDisplayName());
        documentReferenceUser =
                firebaseFirestore.collection("Objects").document(firebaseUser.getDisplayName());
        final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_market_you, null);
        numberPicker = view.findViewById(R.id.numberPicker);
        numberPicker.setMinValue(1);
        numberPicker.setMaxValue(2);
        //  numberPicker.setWrapSelectorWheel(false);

        numberPicker.setOnScrollListener(
                new NumberPicker.OnScrollListener() {
                    @Override
                    public void onScrollStateChange(NumberPicker numberPicker, int i) {
                        Log.i("myTags", "onScrollStateChange, i: " + i);
                    }
                });

        numberPicker.setOnValueChangedListener(
                new NumberPicker.OnValueChangeListener() {
                    @Override
                    public void onValueChange(NumberPicker numberPicker, int i, int i1) {
                        Log.i("myTags", "onValueChange, i1: " + i1);
                    }
                });
        builder.setView(view)
                .setTitle("Sell Sectors")
                .setPositiveButton(
                        "OK",
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                firebaseFirestore.runTransaction(
                                        new Transaction.Function<Void>() {

                                            @Override
                                            public Void apply(Transaction transaction)
                                                    throws FirebaseFirestoreException {
                                                int selectedValue = numberPicker.getValue();
                                                DocumentSnapshot documentSnapshot1 =
                                                        transaction.get(documentReferenceUser);
                                                User user = new User();
                                                user.setMoveToObjectName(
                                                        documentSnapshot1.getString(
                                                                "moveToObjectName"));
                                                user.setMoney(
                                                        documentSnapshot1
                                                                .getLong("money")
                                                                .intValue());
                                                Log.i(
                                                        "myTags",
                                                        "Planet Dialog - apply: current money: "
                                                                + user.getMoney());

                                                DocumentSnapshot documentSnapshot =
                                                        transaction.get(documentReferenceInventory);
                                                user.setSectors(
                                                        documentSnapshot
                                                                .getLong(user.getMoveToObjectName())
                                                                .intValue());

                                                DocumentReference documentReferencePlanet =
                                                        firebaseFirestore
                                                                .collection("Objects")
                                                                .document(
                                                                        user.getMoveToObjectName());
                                                DocumentSnapshot documentSnapshotPlanet =
                                                        transaction.get(documentReferencePlanet);
                                                ObjectModel objectModel = new ObjectModel();
                                                objectModel.setPlanetSectorsPrice(
                                                        documentSnapshotPlanet
                                                                .getLong("sectors_price")
                                                                .intValue());
                                                objectModel.setPlanetSectors(
                                                        documentSnapshotPlanet
                                                                .getLong("sectors")
                                                                .intValue());
                                                if (selectedValue == 1) {
                                                    transaction.update(
                                                            documentReferenceInventory,
                                                            user.getMoveToObjectName(),
                                                            1);
                                                } else {
                                                    transaction.update(
                                                            documentReferenceInventory,
                                                            user.getMoveToObjectName(),
                                                            0);
                                                }
                                                transaction.update(
                                                        documentReferenceUser,
                                                        "money",
                                                        user.getMoney()
                                                                + objectModel
                                                                                .getPlanetSectorsPrice()
                                                                        / selectedValue);
                                                transaction.update(
                                                        documentReferencePlanet,
                                                        "sectors",
                                                        objectModel.getPlanetSectors()
                                                                + selectedValue);
                                                dismiss();
                                                return null;
                                            }
                                        });
                            }
                        });
        return builder.create();
    }
}

File: app/src/main/java/com/artamonov/millionplanets/sectors/SectorsYouDialog.kt
package com.artamonov.millionplanets.sectors

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatDialogFragment
import com.artamonov.millionplanets.R
import com.artamonov.millionplanets.model.SpaceObject
import com.artamonov.millionplanets.model.User
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.android.synthetic.main.dialog_market_you.*
import javax.inject.Inject

@AndroidEntryPoint
class SectorsYouDialog : AppCompatDialogFragment() {
    private var documentReferenceUser: DocumentReference? = null
    private var documentReferenceInventory: DocumentReference? = null

    @Inject lateinit var firebaseUser: FirebaseUser
    @Inject lateinit var firebaseFirestore: FirebaseFirestore
    @Inject lateinit var userDocument: FirebaseUser

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        documentReferenceInventory = firebaseFirestore.collection("Inventory").document(firebaseUser.displayName!!)
        val builder = AlertDialog.Builder(activity)
        val inflater = requireActivity().layoutInflater
        val view = inflater.inflate(R.layout.dialog_market_you, null)
        numberPicker.apply {
            minValue = 1
            maxValue = 2
            setOnScrollListener { _, i -> Log.i("myTags", "onScrollStateChange, i: $i") }
            setOnValueChangedListener { _, _, i1 -> Log.i("myTags", "onValueChange, i1: $i1") }
        }

        builder.setView(view)
                .setTitle("Sell Sectors")
                .setPositiveButton(
                        "OK"
                ) { _, i ->
                    firebaseFirestore.runTransaction<Void> { transaction ->
                        val selectedValue = numberPicker?.value
                        val documentSnapshot1 = transaction[documentReferenceUser!!]
                        val user = User()
                        user.locationName = documentSnapshot1.getString(
                                "locationName")
                        user.money = documentSnapshot1
                                .getLong("money")
                        Log.i(
                                "myTags", "Planet Dialog - apply: current money: " +
                                user.money)
                        val documentSnapshot = transaction[documentReferenceInventory!!]
//                        user.sectors = documentSnapshot
//                                .getLong(user.locationName!!)
                        val documentReferencePlanet = firebaseFirestore
                                .collection("Objects")
                                .document(user.locationName!!)
                        val documentSnapshotPlanet = transaction[documentReferencePlanet]
                        val spaceObject = SpaceObject()
                        spaceObject.planetSectorsPrice = documentSnapshotPlanet
                                .getLong("sectors_price")!!
                        spaceObject.availableSectors = documentSnapshotPlanet
                                .getLong("sectors")!!
                        if (selectedValue == 1) {
                            transaction.update(
                                    documentReferenceInventory!!,
                                    user.locationName!!,
                                    1)
                        } else {
                            transaction.update(
                                    documentReferenceInventory!!,
                                    user.locationName!!,
                                    0)
                        }
                        transaction.update(
                                documentReferenceUser!!,
                                "money", user.money!! +
                                spaceObject.planetSectorsPrice /
                                selectedValue!!)
                        transaction.update(
                                documentReferencePlanet,
                                "sectors", spaceObject.availableSectors +
                                selectedValue)
                        dismiss()
                        null
                    }
                }
        return builder.create()
    }
}